src/http/http_client.f90

```diff
@@ -119,33 +123,52 @@ & function failed. This can occur due to insufficient memory available in the sy
         ! setting response status_code
         rc = curl_easy_getinfo(curl_ptr, CURLINFO_RESPONSE_CODE, response%status_code)  
         
-        ! populate the response header_key array
-        call  response%header%set_header_key()
-        
         call curl_easy_cleanup(curl_ptr)
       
     end function client_get_response
 
-    subroutine prepare_request_header_ptr(header_list_ptr, request)
-        class(request_type), intent(in) :: request
+    subroutine prepare_request_header_ptr(header_list_ptr, req_headers)
         type(c_ptr), intent(out) :: header_list_ptr
-        type(string_type), allocatable :: req_headers(:)
+        type(header_type), allocatable, intent(in) :: req_headers(:)
         character(:), allocatable :: h_key, h_val, final_header_string
         integer :: i
 
-        req_headers = request%header%keys()
         do i = 1, size(req_headers)
-            h_key = char(req_headers(i))
-            h_val = request%header%value(req_headers(i))
-
+            h_key = req_headers(i)%key
+            h_val = req_headers(i)%value
             final_header_string = h_key // ':' // h_val // c_null_char
             header_list_ptr = curl_slist_append(header_list_ptr, final_header_string)
         end do
     end subroutine prepare_request_header_ptr
 
     subroutine set_default_headers(header)
-        type(header_type), intent(inout) :: header
-        call header%set('User-Agent', 'fortran-http/1.0.0')
+        type(header_type), allocatable, intent(inout) :: header(:)
+        type(header_type), allocatable :: default_header(:)
+        integer :: i, j
+        logical :: found
+
+        ! setting all default header
+        default_header = [header_type('user-agent', 'fortran-http/1.0.0')]
+        
+        do j = 1, size(default_header)
+            found = .false.
+            if(allocated(header)) then
+                
+                ! find if default headers are already set by user or not
+                do i = 1, size(header)
+                    if(to_lower(string_type(header(i)%key)) == default_header(j)%key) then
+                        found = .true.
+                        exit
+                    end if
+                end do   
+                
+                if(.not. found) then
+                    call append_header(header, default_header(j)%key, default_header(j)%value)
+                end if
+            else 
+                call append_header(header, default_header(j)%key, default_header(j)%value)
```

Comment:
I don't think you need a separate `append_header` here. In Fortran you can do just:

```fortran
header = [header, header_type(default_header(j)%key, default_header(j)%value)]
```

This of-course re-allocates the entire `header` (as you do in the `append_header` subroutine), but that's fine since we expect most requests and responses to have O(10) headers so the performance hit from re-allocation is not a concern here.